Exercise: Data Pre-processing
- Name of activity: Data Pre-processing
- Your name: Dima Mikhaylov
- Your UVA computing ID: agp7dp

In [45]:
# Imports and dependencies 
import pandas as pd
import numpy as np

#### Step 1: load and inspect the dataset 

Source: https://www.fdic.gov/resources/resolutions/bank-failures/failed-bank-list/
The data was retrieved on 10/22/2021 using download functionality.

In [71]:
df = pd.read_csv('failed_banks.csv')
df.head()

CERT CHCLASS1       CITYST     COST    FAILDATE    FIN    ID  \
0  21111        N   NEWARK, NJ   2491.0  11/01/2019  10534  4100   
1  58112       NM   LOUISA, KY   4547.0  10/25/2019  10532  4098   
2  58317       SB   MAUMEE, OH   2188.0  10/25/2019  10533  4099   
3  10716       NM   COOPER, TX  21577.0  05/31/2019  10531  4097   
4  30570       SB  CHICAGO, IL  86826.0  12/15/2017  10530  4096   

                                  NAME  QBFASSET  QBFDEP  RESTYPE RESTYPE1  \
0     CITY NATIONAL BANK OF NEW JERSEY    120574  111234  FAILURE       PA   
1                LOUISA COMMUNITY BANK     28163   25174  FAILURE       PA   
2                        RESOLUTE BANK     23292   22885  FAILURE       PA   
3                     ENLOE STATE BANK     36738   31254  FAILURE       PI   
4  WASHINGTON FEDERAL BANK FOR SAVINGS    166345  143964  FAILURE       PI   

  SAVR  
0  DIF  
1  DIF  
2  DIF  
3  DIF  
4  DIF

In [72]:
df.shape

(577, 13)

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 577 entries, 0 to 576
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   CERT      577 non-null    int64  
 1   CHCLASS1  577 non-null    object 
 2   CITYST    577 non-null    object 
 3   COST      563 non-null    float64
 4   FAILDATE  577 non-null    object 
 5   FIN       577 non-null    int64  
 6   ID        577 non-null    int64  
 7   NAME      577 non-null    object 
 8   QBFASSET  577 non-null    int64  
 9   QBFDEP    577 non-null    int64  
 10  RESTYPE   577 non-null    object 
 11  RESTYPE1  577 non-null    object 
 12  SAVR      577 non-null    object 
dtypes: float64(1), int64(5), object(7)
memory usage: 58.7+ KB


In [74]:
df.describe()

CERT          COST           FIN           ID      QBFASSET  \
count    577.000000  5.630000e+02    577.000000   577.000000  5.770000e+02   
mean   31539.859619  1.302998e+05   9762.890815  3812.000000  6.817803e+06   
std    16458.879911  5.876533e+05   1908.077978   166.709828  8.064436e+07   
min       91.000000  0.000000e+00      0.000000  3524.000000  3.785000e+03   
25%    20115.000000  1.480600e+04  10102.000000  3668.000000  9.086700e+04   
50%    32102.000000  3.833400e+04  10246.000000  3812.000000  2.039550e+05   
75%    35356.000000  9.262400e+04  10390.000000  3956.000000  5.048160e+05   
max    58701.000000  1.207221e+07  10534.000000  4100.000000  1.471631e+09   

             QBFDEP  
count  5.770000e+02  
mean   3.274679e+06  
std    4.172512e+07  
min    5.000000e+02  
25%    7.854900e+04  
50%    1.827940e+05  
75%    4.434810e+05  
max    9.546776e+08

In [75]:
df.isnull().sum().sum()

14

In [76]:
missing_data = df[df.isnull().any(axis=1)]
missing_data

CERT CHCLASS1              CITYST  COST    FAILDATE   FIN    ID  \
505   3510        N       CHARLOTTE, NC   NaN  01/16/2009     0  3586   
507  25178        N   SAN FRANCISCO, CA   NaN  01/16/2009     0  3588   
508  27374       NM  SALT LAKE CITY, UT   NaN  01/16/2009     0  3589   
509  33143       SB      ALEXANDRIA, VA   NaN  01/16/2009     0  3590   
510  33318        N      WILMINGTON, DE   NaN  01/16/2009     0  3591   
511  34571       SB        NEW YORK, NY   NaN  01/16/2009     0  3592   
513  35453        N        PORTLAND, OR   NaN  01/16/2009     0  3594   
514  58032        N      PROVIDENCE, RI   NaN  01/16/2009     0  3595   
518   7213        N       LAS VEGAS, NV   NaN  11/23/2008     0  3578   
519  18708       SB      WILMINGTON, DE   NaN  11/23/2008     0  3579   
520  18923       NM    CENTURY CITY, CA   NaN  11/23/2008     0  3580   
521  23360        N     SIOUX FALLS, SD   NaN  11/23/2008     0  3581   
522  58180        N     SIOUX FALLS, SD   NaN  11/23/2008     0  3582   
561  26652        N      BOCA RATON, FL   NaN  03/01/2002  4654  3539   

                                  NAME    QBFASSET     QBFDEP     RESTYPE  \
505               BANK OF AMERICA N.A.  1471631047  954677580  ASSISTANCE   
507    BANK OF AMERICA CALIFORNIA N.A.    21502223        500  ASSISTANCE   
508             MERRILL LYNCH BANK USA    61809503   53620324  ASSISTANCE   
509               COUNTRYWIDE BANK FSB   117978966   43663251  ASSISTANCE   
510             FIA CARD SERVICES N.A.   159637136    9816599  ASSISTANCE   
511  MERRILL LYNCH BANK & TRUST CO FSB    37986413   28539028  ASSISTANCE   
513        BANK OF AMERICA OREGON N.A.    11526309        500  ASSISTANCE   
514  BANK OF AMERICA RHODE ISLAND N.A.    35410586        500  ASSISTANCE   
518     CITIBANK, NATIONAL ASSOCIATION  1207007000  230042000  ASSISTANCE   
519           CITICORP TRUST BANK, FSB    19599414    7231013  ASSISTANCE   
520             CITIBANK (BANAMEX USA)     1322720     876086  ASSISTANCE   
521      CITIBANK (SOUTH DAKOTA), N.A.    77737957   42356986  ASSISTANCE   
522    DEPARTMENT STORES NATIONAL BANK      374903     300881  ASSISTANCE   
561            NET FIRST NATIONAL BANK       37207      31601     FAILURE   

    RESTYPE1 SAVR  
505      A/A  ***  
507      A/A  ***  
508      A/A  ***  
509      A/A  ***  
510      A/A  ***  
511      A/A  ***  
513      A/A  ***  
514      A/A  ***  
518      A/A  ***  
519      A/A  ***  
520      A/A  ***  
521      A/A  ***  
522      A/A  ***  
561       PI  BIF

#### It appears that 14 failed banks have `NaN` for costs. These missing values can be imputed either with zeros, or column average, or, even better, average % of costs to assets, as bailing out a bigger bank is more expensive. Need to check how these missing values related to the type pf resolution, `RESTYPE`. As shown below, there are 2 types - `ASSISTANCE` and `FAILURE`.

In [77]:
df.groupby([df['RESTYPE']]).count()

CERT  CHCLASS1  CITYST  COST  FAILDATE  FIN   ID  NAME  QBFASSET  \
RESTYPE                                                                        
ASSISTANCE    13        13      13     0        13   13   13    13        13   
FAILURE      564       564     564   563       564  564  564   564       564   

            QBFDEP  RESTYPE1  SAVR  
RESTYPE                             
ASSISTANCE      13        13    13  
FAILURE        564       564   564

#### Comparing grouping with missing values. It appears that all 13 `ASSISTANCE` cases have `NaN` for `COST` and thus it is warranted to impute these values with zeros. There is only 1 ("NET FIRST NATIONAL BANK") missing value in case of an actual `FAILURE` and it should probably be imputed with some kind of proxy for `COST`.

In [79]:
missing_data.groupby([missing_data['RESTYPE']]).count()

CERT  CHCLASS1  CITYST  COST  FAILDATE  FIN  ID  NAME  QBFASSET  \
RESTYPE                                                                       
ASSISTANCE    13        13      13     0        13   13  13    13        13   
FAILURE        1         1       1     0         1    1   1     1         1   

            QBFDEP  RESTYPE1  SAVR  
RESTYPE                             
ASSISTANCE      13        13    13  
FAILURE          1         1     1

#### Average `COST` of restructuring in case of failure come out to be around 130 million what is probably too high for smaller "NET FIRST NATIONAL BANK", ID=3539 bank with total assets of only 37 million. 

In [97]:
df.loc[df['RESTYPE']=='FAILURE']['COST'].mean()

130299.83836589698

#### Using relative costs seems warranted in this case - need to compute mean of `COST` / `QBFASSET` and apply it to the bank's assets to estimate possible restructuring costs. The overall average is around 23%.

In [107]:
avg_CostsToAssets  = np.round(np.mean(df['COST']/df['QBFASSET']), 2)
avg_CostsToAssets

0.23

#### Applying this `avg_CostsToAssets` multiplier to the bank's assets produced estimated costs of 7.3 million.

In [134]:
df.iloc[561, 3] = df.iloc[561, 9] * avg_CostsToAssets
df.iloc[561, ]

CERT                          26652
CHCLASS1                          N
CITYST               BOCA RATON, FL
COST                        7268.23
FAILDATE                 03/01/2002
FIN                            4654
ID                             3539
NAME        NET FIRST NATIONAL BANK
QBFASSET                      37207
QBFDEP                        31601
RESTYPE                     FAILURE
RESTYPE1                         PI
SAVR                            BIF
Name: 561, dtype: object

#### After this imputation all remaining 13 missing values are associated with `RESTYPE` of assistance that should have zero `COST`. Now one can impute missing values with zeros.

In [138]:
df[df.isnull().any(axis=1)].groupby("RESTYPE").count()

CERT  CHCLASS1  CITYST  COST  FAILDATE  FIN  ID  NAME  QBFASSET  \
RESTYPE                                                                       
ASSISTANCE    13        13      13     0        13   13  13    13        13   

            QBFDEP  RESTYPE1  SAVR  
RESTYPE                             
ASSISTANCE      13        13    13

In [140]:
df['COST'].replace(np.nan, 0, inplace=True)

#### The resulting DataFrame `df` does not have any missing values. 

In [143]:
df.isnull().sum().sum()

0